Fix compilation on Windows host

ppescher opened the pull request:
The native target "mpy-cross" is missing the ".exe" filename extension on Windows.
Building the default STM32 port fails with a truncated command line during the link phase, due to the long list of object files in the stm32lib.
The proposed fix writes the linker input files to a response file. An alternative might be using a static library archive.

Tools used for Windows build:
* TDM-GCC (5.1.0 - MINGW64)
* GNU Arm Embedded Toolchain (7-2018-q2-update - 32bit)
* Git for Windows (2.1.7 - MINGW64)

ppescher committed:
mpy-cross: Fix build target filename extension on Windows

ppescher committed:
stm32: Fix too long command line in linker invocation

Building on Windows host may truncate command line and fail.
Instead use a response file as input for the linker.

ppescher committed:
stm32: Fix compiler library path that may contain spaces

Default ARM GCC installer on Windows is using a path that contains spaces (under Program Files).
When passed as a command line argument to the linker it would cause the build to fail.

ppescher committed:
stm32: Fix continuos integration build

Move linker response file from source to build directory

ppescher commented:
It seems arm-none-eabi-ld version used in Travis-CI does not support reading command line arguments from a file. I don't know how to fix this in some other way.

ppescher committed:
stm32: Fix Travis-CI build

Work-around for Windows command line size limit was not supported on old ARM GNU Linker version used by Travis-CI.
Use it only on a Windows build machine.

ppescher commented:
I just limited the patch to builds on a Windows machine only, leaving the Linux/MacOSX builds unaffected.
That should fix Travis-CI results.

dpgeorge commented:
Thanks for the contribution.  Adding .exe to mpy-cross has actually been discussed and attempted before and it's not easy to do correctly.  See #3361.  In particular see this commit: 02b4b23319a84d5b96b9394d8c379ef0549fd208

chrismas9 commented:
The .exe fix is also required for the Windows port. I found just adding .exe causes other anomalies like the map file being name.exe.map

I believe I have a fully working solution for the .exe problem based on what I submitted in #3361. However I need to rebase it and do extensive testing before submitting. I will try to this in the next few weeks. I will also look at the linker issue.

ppescher commented:
I understand this is more of a work-around than a proper solution, especially the linker stuff, but it seems to do its job.
"mpy-cross" is built using the native gcc, but I'm not sure why it may be built with mingw cross-gcc on linux, as it would be used during the build on the native machine.
Anyway the conditional code in the Makefile does already check for the target machine.

By the way, I have checked the Travis-CI build logs and it seems the windows port does indeed build a "micropython.exe" using cross-gcc (no fix required for the file extension, it is already exe), but "mpy-cross" is built for the host linux platform (not sure why).
Maybe another test should be added that also cross-compiles "mpy-cross.exe" for a Windows target. Is a native linux mpy-cross required for the Windows port?

The Makefile in ports/windows does define `PROG = micropython.exe` so I thought it was fine to do the same for mpy-cross.
I do not see how this can lead to having a double ".exe" extension, as reported in #3361.
As for "mpy-cross.exe.map" I think we can live with it, but I guess it can be fixed if you want.

ppescher committed:
mpy-cross: Fix linker map file filename for Windows target

The map file generated by the linker was having a double ".exe.map" filename extension for Windows builds.

dpgeorge commented:
I guess this is still an issue, with the .exe extension.  See related PRs: #4996, #5727

dpgeorge commented:
Closing in favour of #4996.
